File: 1101.rb
# AOJ 1101 A Simple Offline Text Editor

class TextEditor
  def initialize(text)
    @text = text
    @cursor = 0
  end

  def to_s
    text_before_cursor + "^" + text_after_cursor
  end

  def inspect
    %Q|#<TextEditor:0x#{"%x".%(object_id)} "#{to_s}">|
  end

  def forward_char
    @cursor += 1 if @cursor < @text.length
    self
  end

  def forward_word
    if @cursor < @text.length
      m = /^ +/.match(text_after_cursor)
      @cursor += m[0].length if m
    end

    if @cursor < @text.length
      m = /^[^ ]+/.match(text_after_cursor)
      @cursor += m[0].length if m
    end

    self
  end

  def backward_char
    @cursor -= 1 if @cursor > 0
    self
  end

  def backward_word
    if @cursor > 0
      last_blank_pos = text_before_cursor.sub(/\s+$/, "").rindex(" ")
      if last_blank_pos
        @cursor = last_blank_pos + 1
      else
        @cursor = 0
      end
    end

    self
  end

  def insert(text)
    @text = text_before_cursor + text + text_after_cursor
    @cursor += text.length

    self
  end

  def delete_char
    unless @cursor == @text.length
      @text = text_before_cursor + text_after_cursor[1..-1]
    end

    self
  end

  def delete_word
    if @cursor < @text.length
      @text = text_before_cursor +
        text_after_cursor.sub(/ *[^ ]+/, "")
    end

    self
  end

  private
  def text_before_cursor
    if @cursor.zero?
      ""
    else
      @text[0..(@cursor - 1)]
    end
  end

  def text_after_cursor
    if @cursor == @text.length
      ""
    else
      @text[@cursor..-1]
    end
  end
end

n_texts = gets.to_i
n_texts.times do
  t = TextEditor.new(gets.chomp)

  n_commands = gets.to_i
  n_commands.times do
    command, argument = gets.chomp.split(" ", 2)
    argument.sub!(/^ +/, "")

    case command
    when "forward"
      case argument
      when "char"
        t.forward_char
      when "word"
        t.forward_word
      end
    when "backward"
      case argument
      when "char"
        t.backward_char
      when "word"
        t.backward_word
      end
    when "insert"
      m = /^"(.*)"/.match(argument)
      t.insert m[1] if m
    when "delete"
      case argument
      when "char"
        t.delete_char
      when "word"
        t.delete_word
      end
    end
  end

  puts t.to_s
end

File: 1101.cpp
/* AOJ 1101 A Simple Offline Text Editor */

#include <iostream>
#include <string>

using namespace std;

int
main()
{
    int n_lines;

    cin >> n_lines;
    cin.ignore();

    while (n_lines--) {
        string text;
        int cursor = 0;
        int n_commands;

        getline(cin, text);
        cin >> n_commands;
        cin.ignore();

        while (n_commands--) {
            string command;
            getline(cin, command);

            if (command == "forward char") {
                if (cursor < text.size()) cursor++;
            } else if (command == "forward word") {
                while (cursor < text.size() && text[cursor] == ' ') cursor++;
                while (cursor < text.size() && text[cursor] != ' ') cursor++;
            } else if (command == "backward char") {
                if (cursor > 0) cursor--;
            } else if (command == "backward word") {
                while (cursor > 0 && text[cursor - 1] == ' ') cursor--;
                while (cursor > 0 && text[cursor - 1] != ' ') cursor--;
            } else if (command.find("insert ") == 0) {
                size_t start = command.find_first_of('"') + 1;
                size_t end = command.find_first_of('"', start);
                string arg = command.substr(start, end - start);

                text = text.substr(0, cursor) + arg + text.substr(cursor);
                cursor += arg.size();
            } else if (command == "delete char") {
                if (cursor < text.size()) {
                    text.erase(cursor, 1);
                }
            } else if (command == "delete word") {
                if (cursor < text.size()) {
                    size_t word_end = cursor;

                    while (word_end < text.size() && text[word_end] == ' ')
                        word_end++;
                    if (word_end == text.size()) continue;

                    while (word_end < text.size() && text[word_end] != ' ')
                        word_end++;

                    text = text.substr(0, cursor) + text.substr(word_end);
                }
            }
        }

        cout << text.substr(0, cursor) << '^' << text.substr(cursor) << endl;
    }
}
